File: Engine/Sources/CardCopilotEngine/Engine/PurchaseRouteAdvisor.swift
import Foundation

/// Confidence/provenance for a purchase route. Route facts are deliberately distinct from
/// issuer-confirmed card contracts: gift-card inventory and merchant behaviour can vary by
/// location, so the UI must not present a community-observed route as guaranteed.
public enum PurchaseRouteEvidenceLevel: String, Codable, Equatable, Sendable {
    case retailerConfirmed
    case communityObserved
    case experimental
}

/// Whether the alternate route still charges the destination purchase itself to a credit card.
/// This is a funding-path fact, not an insurance conclusion. The protection advisor separately
/// decides whether card-linked coverage is relevant for the owner's purchase.
public enum PurchaseRouteProtectionEffect: String, Codable, Equatable, Sendable {
    case preservesDestinationCardCharge
    case replacesDestinationCardCharge
}

/// Final route verdict remains multi-attribute: a route can clearly win on rewards while creating
/// a protection trade-off. We do not collapse that conflict into an invented CAD insurance value.
public enum PurchaseRouteVerdict: String, Equatable, Sendable {
    case rewardAdvantage
    case rewardProtectionTradeoff
}

/// A known alternative way to fund a purchase. The route does not hard-code a card: the existing
/// `RecommendationEngine` re-scores the acquisition leg against the owner's actual wallet.
public struct AlternativePurchaseRoute: Codable, Equatable, Sendable, Identifiable {
    public let routeId: String
    public let destinationMerchantAliases: [String]
    public let instrumentLabel: String
    public let acquisitionMerchantLabel: String
    public let acquisitionCategory: String
    public let acquisitionMcc: Int?
    public let acquisitionMerchantBrand: String?
    public let acceptedNetworks: Set<Network>
    public let fixedFeeCad: Double
    public let estimatedFrictionCad: Double
    public let evidenceLevel: PurchaseRouteEvidenceLevel
    public let protectionEffect: PurchaseRouteProtectionEffect
    public let disclosure: String

    public var id: String { routeId }

    public init(routeId: String,
                destinationMerchantAliases: [String],
                instrumentLabel: String,
                acquisitionMerchantLabel: String,
                acquisitionCategory: String,
                acquisitionMcc: Int? = nil,
                acquisitionMerchantBrand: String? = nil,
                acceptedNetworks: Set<Network> = [.amex, .visa, .mastercard],
                fixedFeeCad: Double = 0,
                estimatedFrictionCad: Double = 0,
                evidenceLevel: PurchaseRouteEvidenceLevel,
                protectionEffect: PurchaseRouteProtectionEffect = .preservesDestinationCardCharge,
                disclosure: String) {
        self.routeId = routeId
        self.destinationMerchantAliases = destinationMerchantAliases
        self.instrumentLabel = instrumentLabel
        self.acquisitionMerchantLabel = acquisitionMerchantLabel
        self.acquisitionCategory = CategoryTaxonomy.canonicalID(acquisitionCategory)
        self.acquisitionMcc = acquisitionMcc
        self.acquisitionMerchantBrand = acquisitionMerchantBrand
        self.acceptedNetworks = acceptedNetworks
        self.fixedFeeCad = fixedFeeCad
        self.estimatedFrictionCad = estimatedFrictionCad
        self.evidenceLevel = evidenceLevel
        self.protectionEffect = protectionEffect
        self.disclosure = disclosure
    }

    public func matches(destinationMerchantName: String) -> Bool {
        let merchant = Self.normalized(destinationMerchantName)
        return destinationMerchantAliases.contains { alias in
            let candidate = Self.normalized(alias)
            return !candidate.isEmpty && merchant.contains(candidate)
        }
    }

    private static func normalized(_ value: String) -> String {
        value.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }
}

/// Separate from the card-switch threshold: an alternate route can require another stop or an
/// extra payment instrument, so it needs a materially higher bar than merely pulling another card.
public struct PurchaseRouteThreshold: Equatable, Sendable {
    public let minAdvantageCad: Double
    public let minAdvantagePercentagePoints: Double

    public init(minAdvantageCad: Double = 1,
                minAdvantagePercentagePoints: Double = 1) {
        self.minAdvantageCad = minAdvantageCad
        self.minAdvantagePercentagePoints = minAdvantagePercentagePoints
    }

    public static let shipped = PurchaseRouteThreshold()
}

public struct PurchaseRouteEvaluation: Equatable, Sendable {
    public let route: AlternativePurchaseRoute
    public let acquisitionRecommendation: Recommendation
    /// Reward/credit decision value only. Protection stays separate in `protectionAssessment`.
    public let directValueCad: Double
    public let routeValueCad: Double
    public let advantageCad: Double
    public let advantagePercentagePoints: Double
    public let protectionAssessment: ProtectionDecisionAssessment
    public let verdict: PurchaseRouteVerdict
}

/// Compares the best direct card with known alternate funding paths. This sits above
/// `RecommendationEngine`: it changes *how* the purchase is made, while the existing engine stays
/// the sole authority for which owned card wins each card-funded leg.
public enum PurchaseRouteAdvisor {
    public static func bestAlternative(
        directRecommendation: Recommendation,
        destination: PurchaseContext,
        destinationMerchantName: String,
        routes: [AlternativePurchaseRoute] = PurchaseRouteCatalogue.canadaV1,
        engine: RecommendationEngine,
        asOf: String,
        threshold: PurchaseRouteThreshold = .shipped,
        benefits: BenefitsCatalogue? = nil,
        declaredBenefitContext: BenefitContext? = nil
    ) -> PurchaseRouteEvaluation? {
        guard destination.amountCad > 0 else { return nil }

        let directValue = directRecommendation.winner.decisionValueCad
        var eligible: [PurchaseRouteEvaluation] = []

        for route in routes where route.matches(destinationMerchantName: destinationMerchantName) {
            let acquisition = PurchaseContext(
                amountCad: destination.amountCad,
                currency: destination.currency,
                usdEquivalent: destination.usdEquivalent,
                category: route.acquisitionCategory,
                mcc: route.acquisitionMcc,
                merchantBrand: route.acquisitionMerchantBrand,
                country: destination.country,
                channel: "cardPresent",
                recurringIndicator: false,
                acceptedNetworks: route.acceptedNetworks
            )

            guard case .advised(let recommendation) = engine.recommend(acquisition, asOf: asOf) else {
                continue
            }

            let routeValue = recommendation.winner.decisionValueCad
                - route.fixedFeeCad
                - route.estimatedFrictionCad
            let advantage = routeValue - directValue
            let advantagePP = advantage / destination.amountCad * 100

            // Require BOTH reward floors. Protection does not add or subtract an invented CAD
            // amount; it changes the final verdict below.
            guard advantage >= threshold.minAdvantageCad,
                  advantagePP >= threshold.minAdvantagePercentagePoints else { continue }

            let protectionAssessment: ProtectionDecisionAssessment
            if route.protectionEffect == .replacesDestinationCardCharge,
               let benefits {
                protectionAssessment = ProtectionDecisionAdvisor.alternateFundingAssessment(
                    directCardId: directRecommendation.winner.cardId,
                    purchase: destination,
                    benefits: benefits,
                    declaredContext: declaredBenefitContext)
            } else {
                protectionAssessment = ProtectionDecisionAssessment(
                    status: .notRelevant,
                    directCardId: directRecommendation.winner.cardId)
            }

            let verdict: PurchaseRouteVerdict = protectionAssessment.status == .notRelevant
                ? .rewardAdvantage
                : .rewardProtectionTradeoff

            eligible.append(PurchaseRouteEvaluation(
                route: route,
                acquisitionRecommendation: recommendation,
                directValueCad: directValue,
                routeValueCad: routeValue,
                advantageCad: advantage,
                advantagePercentagePoints: advantagePP,
                protectionAssessment: protectionAssessment,
                verdict: verdict
            ))
        }

        // Reward ordering remains explicit rather than pretending an unresolved protection
        // trade-off has a dollar price. The UI must render `verdict` beside the reward advantage.
        return eligible.max { lhs, rhs in
            if lhs.advantageCad != rhs.advantageCad {
                return lhs.advantageCad < rhs.advantageCad
            }
            return lhs.route.routeId > rhs.route.routeId
        }
    }
}

/// V1 keeps route knowledge intentionally small. The acquisition merchant is a *requirement*, not
/// a promise that every grocery store stocks the gift card. The merchant/MCC graph can later
/// resolve nearby terminals that satisfy this template without changing route-scoring semantics.
public enum PurchaseRouteCatalogue {
    public static let canadaV1: [AlternativePurchaseRoute] = [
        AlternativePurchaseRoute(
            routeId: "shoppers-gift-card-via-grocery-5411",
            destinationMerchantAliases: ["Shoppers Drug Mart", "Shoppers", "Pharmaprix"],
            instrumentLabel: "Shoppers Drug Mart gift card",
            acquisitionMerchantLabel: "an eligible grocery store that codes as MCC 5411",
            acquisitionCategory: "grocery",
            acquisitionMcc: 5411,
            evidenceLevel: .communityObserved,
            protectionEffect: .replacesDestinationCardCharge,
            disclosure: "Gift-card inventory and issuer reward treatment can vary by store and transaction. Confirm availability before relying on this route."
        )
    ]
}

File: android/core/engine/src/main/kotlin/com/cardcopilot/engine/engine/PurchaseRouteAdvisor.kt
package com.cardcopilot.engine.engine

import com.cardcopilot.engine.models.BenefitsCatalogue
import com.cardcopilot.engine.models.Network
import com.cardcopilot.engine.models.PurchaseContext
import com.cardcopilot.engine.models.Recommendation
import com.cardcopilot.engine.models.RecommendationOutcome

enum class PurchaseRouteEvidenceLevel {
    RETAILER_CONFIRMED,
    COMMUNITY_OBSERVED,
    EXPERIMENTAL
}

enum class PurchaseRouteProtectionEffect {
    PRESERVES_DESTINATION_CARD_CHARGE,
    REPLACES_DESTINATION_CARD_CHARGE
}

enum class PurchaseRouteVerdict {
    REWARD_ADVANTAGE,
    REWARD_PROTECTION_TRADEOFF
}

data class AlternativePurchaseRoute(
    val routeId: String,
    val destinationMerchantAliases: List<String>,
    val instrumentLabel: String,
    val acquisitionMerchantLabel: String,
    val acquisitionCategory: String,
    val acquisitionMcc: Int? = null,
    val acquisitionMerchantBrand: String? = null,
    val acceptedNetworks: Set<Network> = setOf(Network.AMEX, Network.VISA, Network.MASTERCARD),
    val fixedFeeCad: Double = 0.0,
    val estimatedFrictionCad: Double = 0.0,
    val evidenceLevel: PurchaseRouteEvidenceLevel,
    val protectionEffect: PurchaseRouteProtectionEffect = PurchaseRouteProtectionEffect.PRESERVES_DESTINATION_CARD_CHARGE,
    val disclosure: String
) {
    fun matches(destinationMerchantName: String): Boolean {
        val merchant = normalized(destinationMerchantName)
        return destinationMerchantAliases.any { alias ->
            val candidate = normalized(alias)
            candidate.isNotEmpty() && merchant.contains(candidate)
        }
    }

    private fun normalized(value: String): String =
        value.lowercase()
            .replace(Regex("[^a-z0-9]+"), " ")
            .trim()
}

data class PurchaseRouteThreshold(
    val minAdvantageCad: Double = 1.0,
    val minAdvantagePercentagePoints: Double = 1.0
) {
    companion object {
        val SHIPPED = PurchaseRouteThreshold()
    }
}

data class PurchaseRouteEvaluation(
    val route: AlternativePurchaseRoute,
    val acquisitionRecommendation: Recommendation,
    val directValueCad: Double,
    val routeValueCad: Double,
    val advantageCad: Double,
    val advantagePercentagePoints: Double,
    val protectionAssessment: ProtectionDecisionAssessment,
    val verdict: PurchaseRouteVerdict
)

object PurchaseRouteAdvisor {
    fun bestAlternative(
        directRecommendation: Recommendation,
        destination: PurchaseContext,
        destinationMerchantName: String,
        routes: List<AlternativePurchaseRoute> = PurchaseRouteCatalogue.canadaV1,
        engine: RecommendationEngine,
        asOf: String,
        threshold: PurchaseRouteThreshold = PurchaseRouteThreshold.SHIPPED,
        benefits: BenefitsCatalogue? = null,
        declaredBenefitContext: BenefitContext? = null
    ): PurchaseRouteEvaluation? {
        if (destination.amountCad <= 0.0) return null

        val directValue = directRecommendation.winner.decisionValueCad
        val eligible = mutableListOf<PurchaseRouteEvaluation>()

        for (route in routes) {
            if (!route.matches(destinationMerchantName)) continue

            val acquisition = PurchaseContext(
                amountCad = destination.amountCad,
                currency = destination.currency,
                usdEquivalent = destination.usdEquivalent,
                category = route.acquisitionCategory,
                mcc = route.acquisitionMcc,
                merchantBrand = route.acquisitionMerchantBrand,
                country = destination.country,
                channel = "cardPresent",
                recurringIndicator = false,
                acceptedNetworks = route.acceptedNetworks
            )

            val outcome = engine.recommend(acquisition, asOf)
            val recommendation = (outcome as? RecommendationOutcome.Advised)?.recommendation ?: continue
            val routeValue = recommendation.winner.decisionValueCad -
                route.fixedFeeCad - route.estimatedFrictionCad
            val advantage = routeValue - directValue
            val advantagePP = advantage / destination.amountCad * 100.0

            if (advantage < threshold.minAdvantageCad ||
                advantagePP < threshold.minAdvantagePercentagePoints
            ) continue

            val protectionAssessment = if (
                route.protectionEffect == PurchaseRouteProtectionEffect.REPLACES_DESTINATION_CARD_CHARGE &&
                benefits != null
            ) {
                ProtectionDecisionAdvisor.alternateFundingAssessment(
                    directCardId = directRecommendation.winner.cardId,
                    purchase = destination,
                    benefits = benefits,
                    declaredContext = declaredBenefitContext
                )
            } else {
                ProtectionDecisionAssessment(
                    status = ProtectionDecisionStatus.NOT_RELEVANT,
                    directCardId = directRecommendation.winner.cardId
                )
            }
            val verdict = if (protectionAssessment.status == ProtectionDecisionStatus.NOT_RELEVANT) {
                PurchaseRouteVerdict.REWARD_ADVANTAGE
            } else {
                PurchaseRouteVerdict.REWARD_PROTECTION_TRADEOFF
            }

            eligible += PurchaseRouteEvaluation(
                route = route,
                acquisitionRecommendation = recommendation,
                directValueCad = directValue,
                routeValueCad = routeValue,
                advantageCad = advantage,
                advantagePercentagePoints = advantagePP,
                protectionAssessment = protectionAssessment,
                verdict = verdict
            )
        }

        return eligible.maxWithOrNull { lhs, rhs ->
            val advantageOrder = lhs.advantageCad.compareTo(rhs.advantageCad)
            if (advantageOrder != 0) advantageOrder
            else rhs.route.routeId.compareTo(lhs.route.routeId)
        }
    }
}

object PurchaseRouteCatalogue {
    val canadaV1: List<AlternativePurchaseRoute> = listOf(
        AlternativePurchaseRoute(
            routeId = "shoppers-gift-card-via-grocery-5411",
            destinationMerchantAliases = listOf("Shoppers Drug Mart", "Shoppers", "Pharmaprix"),
            instrumentLabel = "Shoppers Drug Mart gift card",
            acquisitionMerchantLabel = "an eligible grocery store that codes as MCC 5411",
            acquisitionCategory = "grocery",
            acquisitionMcc = 5411,
            evidenceLevel = PurchaseRouteEvidenceLevel.COMMUNITY_OBSERVED,
            protectionEffect = PurchaseRouteProtectionEffect.REPLACES_DESTINATION_CARD_CHARGE,
            disclosure = "Gift-card inventory and issuer reward treatment can vary by store and transaction. Confirm availability before relying on this route."
        )
    )
}
